### DISCUSSION TITLE
Query

### DISCUSSION THREAD

**Discussion by ankushgoel27** [AdnaneKhan/Gato-X#34] [General]
Hi AdnaneKhan,

I am new to github actions vulnerability so pardon me with my queries

I have few questions I hope you can answer since you are very well knowledgable about github actions vulnerabilities.

1) The gato-x tool requires `workflow` permissions on the token to exfilterate secrets, it does that by creating workflows and executing those. But wouldn't that be possible also if the token just have `repo` scope. cant i create a PR or push changes to a branch which will execute my malicious workflow(`exfil secrets`) using `push` or `pull_request` events?

also, same method could be used to takeover self hosted runners where the token has write permissions?

Also, would like to chat with you if you are on discord or any other messaging platform.

**Comment by AdnaneKhan** [comment]
Hi, converted to a discussion as it’s better suited to it.

You are correct. With a repo scope you can get secrets, jump to runners with the exception that you cannot modify workflow files themselves. You must use the existing triggers (push, pr, dispatch, etc). Gato-X doesn’t automate it because it is custom exploitation and very specific to each repository.

Would recommend checking out  https://github.com/adnanekhan/conferencetalks and https://github.com/jstawinski/GitHub-Actions-Attack-Diagram as good resources on privesc and lateral movement in actions.

**Reply by ankushgoel27** [reply]
with only repo scope, you have write permission on the repository and the repository is configured to trigger workflows on pull_request/PUSH events, your modified workflow should automatically trigger when you submit the PR or push to a branch. isn't that right? so no workflow scope is required and my modified workflow will run.

Just curious, cant this be automated

**Reply by AdnaneKhan** [reply]
> with only repo scope, you have write permission on the repository and the repository is configured to trigger workflows on pull_request/PUSH events, your modified workflow should automatically trigger when you submit the PR or push to a branch. isn't that right? so no workflow scope is required and my modified workflow will run.
> 
> Just curious, cant this be automated

With a repo scoped token you cannot commit any changes to the .github/workflows directory using git or the API (if you find a direct bypass then that’s a GitHub vulnerability).

**Reply by ankushgoel27** [reply]
> > with only repo scope, you have write permission on the repository and the repository is configured to trigger workflows on pull_request/PUSH events, your modified workflow should automatically trigger when you submit the PR or push to a branch. isn't that right? so no workflow scope is required and my modified workflow will run.
> 
> > 
> 
> > Just curious, cant this be automated
> 
> 
> 
> With a repo scoped token you cannot commit any changes to the .github/workflows directory using git or the API (if you find a direct bypass then that’s a GitHub vulnerability).

Yes, can't commit but don't need to commit. Only create a PR, and it will run the modified workflow file if there is a pull_request trigger. 

Same with your read permission outside collaborator PR which runs but needs approval and doesn't have access to secrets. However, in this case, no approval needed and access to secrets is there.

**Reply by AdnaneKhan** [reply]
Pull requests from forks (regardless of who makes the PR) do not have access to secrets. Pull requests from feature branches (both branches in same repository) do, but in order to modify a workflow in the PR head branch within a repository you’ll need the workflow scope.

If a workflow file already has uses secrets and calls a script that you can modify (outside of the workflows directory), then you can get secrets by modifying that script or unit test. The problem is that it is hard to automatically identify what file to change, because every repository is different.
